enclave/crypto/openssl/cert.c

```diff
@@ -0,0 +1,405 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include "cert.h"
+#include <ctype.h>
+#include <mbedtls/oid.h> // For mbedtls_oid_get_numeric_string
+#include <openenclave/bits/result.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/internal/cert.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/safecrt.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+#include <openssl/bio.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+#include <openssl/x509.h>
+#include <openssl/x509_vfy.h>
+#include <openssl/x509v3.h>
+#include <string.h>
+
+/*
+ * Value used by the _decode_oid_to_str() function. Although the OID
+ * standard does not limit the depth of an OID definition tree (i.e., the
+ * number of arcs), our implementation only supports a simple decoding
+ * with a limited depth (i.e., decoding into a fixed size string).
+ */
+#define OE_OID_MAX_LENGTH 200
+
+/*
+ * Parse the name string into X509_NAME struct. The format of the string is
+ * "KEY1=VALUE1,KEY2=VALUE2,KEY3=VALUE3...". The implementation is based
+ * on the mbedtls_x509_string_to_names from Mbed TLS.
+ */
+X509_NAME* X509_parse_name(const char* name_string)
+{
+    const char* s = name_string;
+    const char* c = s;
+    const char* end = s + strlen(s);
+    int in_tag = 1;
+    char key[OE_X509_MAX_NAME_SIZE];
+    char value[OE_X509_MAX_NAME_SIZE];
+    char data[OE_X509_MAX_NAME_SIZE];
+    char* d = data;
+    X509_NAME* name = NULL;
+    int error = 1;
+
+    name = X509_NAME_new();
+    if (name == NULL)
+        goto done;
+
+    while (c <= end)
+    {
+        if (in_tag && *c == '=')
+        {
+            size_t len = (size_t)(c - s) + 1;
+            if (len > OE_X509_MAX_NAME_SIZE)
+                goto done;
+
+            if (oe_memcpy_s(key, OE_X509_MAX_NAME_SIZE, s, len) != OE_OK)
+                goto done;
+            key[len - 1] = '\0';
+            s = c + 1;
+            in_tag = 0;
+            d = data;
+        }
+
+        if (!in_tag && *c == '\\' && c != end)
+        {
+            c++;
+            /* Only support escaping commas */
+            if (c == end || *c != ',')
+                goto done;
+        }
+        else if (!in_tag && (*c == ',' || c == end))
+        {
+            /*
+             * The check of if(d - data == OE_X509_MAX_NAME_SIZE)
+             * below ensures that len should never exceed OE_X509_MAX_NAME_SIZE.
+             * We do not perform additional checks here.
+             */
+            size_t len = (size_t)(d - data) + 1;
+            if (oe_memcpy_s(value, OE_X509_MAX_NAME_SIZE, data, len) != OE_OK)
+                goto done;
+            value[len - 1] = '\0';
+            if (!X509_NAME_add_entry_by_txt(
+                    name, key, MBSTRING_ASC, (unsigned char*)value, -1, -1, 0))
```

Comment:
Just in terms of parity with MbedTLS behavior, I don't think it handles multibyte ASCII strings. I can't find documentation on this, but minimally MbedTLS has to handle UTF-8 strings for ASN ... is there a reason we're using `MBSTRING_ASC` here instead of `MBSTRING_UTF8`?

Reply:
Updated to use MBSTRING_UTF8 instead.